### PR TITLE
[WIP] Teach SourceKit name translation request to translate Swift names to ObjC ones (by using PrintAsObjC).

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci smoke test

**Comment by nkcsgexi** [comment]
@swift-ci smoke test

File: include/swift/AST/Identifier.h [hunk 1]
```diff
@@ -433,6 +433,12 @@ class ObjCSelector {
   /// pieces.
   ObjCSelector(ASTContext &ctx, unsigned numArgs, ArrayRef<Identifier> pieces);
 
+  /// Construct an invalid ObjCSelector.
+  ObjCSelector() : Storage() {};
```
Comment (jrose-apple) [inline include/swift/AST/Identifier.h]: Nitpick: trailing semicolon. :-)

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 2]
```diff
@@ -2633,3 +2633,34 @@ bool swift::printAsObjC(llvm::raw_ostream &os, ModuleDecl *M,
...
+      SmallString<64> Scratch;
+      OS << camel_case::toSentencecase(PreferredName.getBaseName().str(),
+                                       Scratch);
+    }
+    return {Ctx.getIdentifier(Buffer.str()), ObjCSelector()};
+  } else {
+    auto Name = getNameForObjC(VD, CustomNamesOnly);
+    if (!Name.empty())
+      return {Ctx.getIdentifier(Name), ObjCSelector()};
+    return {PreferredName.getBaseName(), ObjCSelector()};
```
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Hang on, what if PreferredName is also null? (Classes and protocols would currently fall into this case.)

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 3]
```diff
@@ -877,14 +878,47 @@ getClangDeclarationName(clang::ASTContext &Ctx, NameTranslatingInfo &Info) {
   }
 }
 
+static DeclName
+getSwiftDeclName(ASTContext &Ctx, NameTranslatingInfo &Info) {
+  assert(SwiftLangSupport::getNameKindForUID(Info.NameKind) == NameKind::Swift);
+  std::vector<Identifier> Args(Info.ArgNames.size(), Identifier());
```
Comment (jrose-apple) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp]: Nitpick: Why not SmallVector?

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 4]
```diff
@@ -877,14 +878,47 @@ getClangDeclarationName(clang::ASTContext &Ctx, NameTranslatingInfo &Info) {
...
+      StringRef Total = Selector.getString(Buffer);
+      SmallVector<StringRef, 4> Pieces;
+      Total.split(Pieces, ":");
+      if (Selector.getNumArgs()) {
+        assert(Pieces.back().empty());
+        Pieces.pop_back();
+        std::transform(Pieces.begin(), Pieces.end(), Pieces.begin(),
+          [](StringRef P) { return StringRef(P.data(), P.size() + 1); });
+      }
+      Result.ArgNames = llvm::makeArrayRef(Pieces);
```
Comment (jrose-apple) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp]: This looks suspicious, even though I know ArgNames isn't an ArrayRef itself. Can you go through a different API just to make it not look like you're storing the ArrayRef?

File: tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp [hunk 5]
```diff
@@ -593,7 +593,7 @@ static int handleTestInvocation(ArrayRef<const char *> Args,
       } else {
         BaseName = Text.substr(0, ArgStart);
         auto ArgEnd = Text.find_last_of(')');
-        if (ArgEnd != StringRef::npos) {
+        if (ArgEnd == StringRef::npos) {
```
Comment (jrose-apple) [inline tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp]: Heh, oops.

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 6]
```diff
@@ -2633,3 +2633,34 @@ bool swift::printAsObjC(llvm::raw_ostream &os, ModuleDecl *M,
...
+  } else if (auto *SD = dyn_cast<SubscriptDecl>(VD)) {
+    return getObjCNameForSwiftDecl(SD->getGetter(), PreferredName);
+  } else if (auto *EL = dyn_cast<EnumElementDecl>(VD)) {
+    EnumDecl* ED = EL->getDeclContext()->getAsEnumOrEnumExtensionContext();
+    SmallString<64> Buffer;
+    {
+      llvm::raw_svector_ostream OS(Buffer);
+      OS << getNameForObjC(ED).str();
+      SmallString<64> Scratch;
+      OS << camel_case::toSentencecase(PreferredName.getBaseName().str(),
```
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Can you abstract this out so that the actual printing code can use this too?

**Review by jrose-apple — COMMENTED** [review]
Looks pretty good, but I do have a few comments.

**Comment by nkcsgexi** [comment]
oh, this is already merged. I'll address your comments in another pull request.

File: lib/PrintAsObjC/PrintAsObjC.cpp (hunk 2 above)
Comment (nkcsgexi) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: We get an empty identifier then. I think it is fine.

Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Isn't it supposed to return the current name in that case?

**Pull request merged**